File: src/commands/state.ts
import path from 'path';
import ora from 'ora';
import chalk from 'chalk';
import { FileSystemUtils } from '../utils/file-system.js';
import { getProjectorDirName } from '../core/project-config.js';

export class StateCommand {
  private async getStatePath(projectPath: string): Promise<string> {
    const projectorDir = getProjectorDirName(projectPath);
    return path.join(projectPath, projectorDir, 'planning', 'STATE.md');
  }

  private async ensureStateFile(statePath: string): Promise<void> {
    if (!(await FileSystemUtils.fileExists(statePath))) {
      throw new Error(
        'STATE.md not found. Run "projector init" first or create projector/planning/STATE.md'
      );
    }
  }

  private getCurrentDate(): string {
    return new Date().toISOString().split('T')[0];
  }

  private getCurrentTimestamp(): string {
    return new Date().toISOString().replace('T', ' ').split('.')[0];
  }

  async show(projectPath: string = '.'): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);
    console.log(content);
  }

  async addDecision(
    text: string,
    projectPath: string = '.'
  ): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);
    const date = this.getCurrentDate();
    const entry = `- ${date}: ${text}`;

    // Find the "## Recent Decisions" section and add the entry
    const updatedContent = this.insertAfterSection(
      content,
      '## Recent Decisions',
      entry
    );

    await FileSystemUtils.writeFile(statePath, updatedContent);
    await this.updateLastUpdated(statePath);

    ora().succeed(chalk.green(`Decision recorded: ${text}`));
  }

  async addBlocker(
    text: string,
    projectPath: string = '.'
  ): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);
    const entry = `- ${text}`;

    // Find the "## Blockers" section and add the entry
    let updatedContent = content;

    // Replace "[None currently]" if present
    if (content.includes('[None currently]')) {
      updatedContent = content.replace('[None currently]', entry);
    } else {
      updatedContent = this.insertAfterSection(content, '## Blockers', entry);
    }

    await FileSystemUtils.writeFile(statePath, updatedContent);
    await this.updateLastUpdated(statePath);

    ora().succeed(chalk.green(`Blocker recorded: ${text}`));
  }

  async addNote(
    text: string,
    projectPath: string = '.'
  ): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);
    const date = this.getCurrentDate();
    const timestamp = this.getCurrentTimestamp();

    // Check if there's already a session note for today
    const todayHeader = `### ${date}`;

    let updatedContent: string;
    if (content.includes(todayHeader)) {
      // Add to existing today's session
      const entry = `- ${timestamp.split(' ')[1]}: ${text}`;
      updatedContent = this.insertAfterSection(content, todayHeader, entry);
    } else {
      // Create new session section
      const sessionEntry = `### ${date} Session\n- ${timestamp.split(' ')[1]}: ${text}`;
      updatedContent = this.insertAfterSection(
        content,
        '## Session Notes',
        sessionEntry
      );
    }

    await FileSystemUtils.writeFile(statePath, updatedContent);
    await this.updateLastUpdated(statePath);

    ora().succeed(chalk.green(`Note recorded: ${text}`));
  }

  async setFocus(
    text: string,
    projectPath: string = '.'
  ): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);

    // Replace the current focus section content
    const focusRegex = /(## Current Focus\n)([^\n#]*)/;
    const updatedContent = content.replace(focusRegex, `$1${text}`);

    await FileSystemUtils.writeFile(statePath, updatedContent);
    await this.updateLastUpdated(statePath);

    ora().succeed(chalk.green(`Focus updated: ${text}`));
  }

  async addQuestion(
    text: string,
    projectPath: string = '.'
  ): Promise<void> {
    const resolvedPath = path.resolve(projectPath);
    const statePath = await this.getStatePath(resolvedPath);

    await this.ensureStateFile(statePath);

    const content = await FileSystemUtils.readFile(statePath);
    const entry = `- [ ] ${text}`;

    const updatedContent = this.insertAfterSection(
      content,
      '## Open Questions',
      entry
    );

    await FileSystemUtils.writeFile(statePath, updatedContent);
    await this.updateLastUpdated(statePath);

    ora().succeed(chalk.green(`Question added: ${text}`));
  }

  private insertAfterSection(
    content: string,
    sectionHeader: string,
    entry: string
  ): string {
    const lines = content.split('\n');
    const sectionIndex = lines.findIndex((line) =>
      line.startsWith(sectionHeader)
    );

    if (sectionIndex === -1) {
      // Section not found, append at the end before "---" divider or at the very end
      const dividerIndex = lines.findIndex((line) => line.trim() === '---');
      if (dividerIndex !== -1) {
        lines.splice(dividerIndex, 0, '', sectionHeader, entry);
      } else {
        lines.push('', sectionHeader, entry);
      }
    } else {
      // Find the next section or end, insert before it
      let insertIndex = sectionIndex + 1;

      // Skip empty lines after header
      while (insertIndex < lines.length && lines[insertIndex].trim() === '') {
        insertIndex++;
      }

      // Insert the new entry
      lines.splice(insertIndex, 0, entry);
    }

    return lines.join('\n');
  }

  private async updateLastUpdated(statePath: string): Promise<void> {
    const content = await FileSystemUtils.readFile(statePath);
    const date = this.getCurrentDate();

    const updatedContent = content.replace(
      /Last Updated: .*/,
      `Last Updated: ${date}`
    );

    await FileSystemUtils.writeFile(statePath, updatedContent);
  }
}

File: spool-rs/crates/spool-domain/src/state.rs
use chrono::Local;
use regex::Regex;

pub fn update_last_updated(contents: &str, date: &str) -> String {
    let re = Regex::new(r"(?m)^Last Updated: .+$").unwrap();
    if re.is_match(contents) {
        return re
            .replace_all(contents, format!("Last Updated: {date}").as_str())
            .to_string();
    }
    contents.to_string()
}

fn insert_after_heading(
    contents: &str,
    heading: &str,
    line_to_insert: &str,
) -> Result<String, String> {
    let mut out: Vec<String> = Vec::new();
    let mut inserted = false;
    for line in contents.lines() {
        out.push(line.to_string());
        if !inserted && line.trim() == heading {
            out.push(line_to_insert.to_string());
            inserted = true;
        }
    }
    if !inserted {
        return Err(format!("Missing heading: {heading}"));
    }
    let mut s = out.join("\n");
    s.push('\n');
    Ok(s)
}

pub fn add_decision(contents: &str, date: &str, text: &str) -> Result<String, String> {
    let line = format!("- {date}: {text}");
    let updated = insert_after_heading(contents, "## Recent Decisions", &line)?;
    Ok(update_last_updated(&updated, date))
}

pub fn add_question(contents: &str, date: &str, text: &str) -> Result<String, String> {
    let line = format!("- [ ] {text}");
    let updated = insert_after_heading(contents, "## Open Questions", &line)?;
    Ok(update_last_updated(&updated, date))
}

pub fn add_blocker(contents: &str, date: &str, text: &str) -> Result<String, String> {
    let mut lines: Vec<String> = contents.lines().map(|l| l.to_string()).collect();
    let mut i = 0usize;
    let mut in_blockers = false;
    let mut inserted = false;
    while i < lines.len() {
        let line = lines[i].as_str();
        if line.trim() == "## Blockers" {
            in_blockers = true;
            i += 1;
            continue;
        }
        if in_blockers {
            if line.starts_with("## ") {
                lines.insert(i, format!("- {text}"));
                inserted = true;
                break;
            }
            if line.trim() == "[None currently]" {
                lines[i] = format!("- {text}");
                inserted = true;
                break;
            }
            if line.trim().is_empty() {
                lines.insert(i, format!("- {text}"));
                inserted = true;
                break;
            }
        }
        i += 1;
    }
    if !inserted {
        return Err("Could not find Blockers section".to_string());
    }
    let mut out = lines.join("\n");
    out.push('\n');
    Ok(update_last_updated(&out, date))
}

pub fn set_focus(contents: &str, date: &str, text: &str) -> Result<String, String> {
    // Match TS: /(## Current Focus\n)([^\n#]*)/
    let re = Regex::new(r"(?m)(## Current Focus\n)([^\n#]*)").unwrap();
    if !re.is_match(contents) {
        return Err("Could not find Current Focus section".to_string());
    }
    let updated = re
        .replace(contents, |caps: &regex::Captures<'_>| {
            format!("{}{}\n", &caps[1], text)
        })
        .to_string();
    Ok(update_last_updated(&updated, date))
}

pub fn add_note(contents: &str, date: &str, time: &str, text: &str) -> Result<String, String> {
    let session_header = format!("### {date} Session");
    let entry = format!("- {time}: {text}");

    let mut lines: Vec<String> = contents.lines().map(|l| l.to_string()).collect();

    // If a matching session header exists, insert immediately after it.
    for i in 0..lines.len() {
        if lines[i].trim() == session_header {
            lines.insert(i + 1, entry);
            let mut out = lines.join("\n");
            out.push('\n');
            return Ok(update_last_updated(&out, date));
        }
    }

    // Otherwise, insert after Session Notes heading.
    for i in 0..lines.len() {
        if lines[i].trim() == "## Session Notes" {
            lines.insert(i + 1, session_header);
            lines.insert(i + 2, entry);
            let mut out = lines.join("\n");
            out.push('\n');
            return Ok(update_last_updated(&out, date));
        }
    }

    Err("Could not find Session Notes section".to_string())
}

pub fn now_time() -> String {
    Local::now().format("%H:%M:%S").to_string()
}

pub fn now_date() -> String {
    Local::now().format("%Y-%m-%d").to_string()
}
